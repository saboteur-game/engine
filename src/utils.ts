import { v4 as uuidv4 } from "uuid";

const generateId = () => uuidv4();

// https://stackoverflow.com/a/2450976
function shuffle<T>(originalArray: T[]): T[] {
  const array = [...originalArray];
  let currentIndex = array.length;
  let currentValue;
  let randomIndex;

  while (0 !== currentIndex) {
    randomIndex = Math.floor(Math.random() * currentIndex);
    currentIndex -= 1;
    currentValue = array[currentIndex];
    array[currentIndex] = array[randomIndex];
    array[randomIndex] = currentValue;
  }

  return array;
}

interface Pojo {
  [key: string]: unknown;
}

export { generateId, shuffle, Pojo };
