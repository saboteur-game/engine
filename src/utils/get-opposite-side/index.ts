import { Sides } from "../../models/cards/card";
import { SPACE_SIDES } from "../../constants";

export const getOppositeSide = (side: Sides): Sides => {
  if (!Sides[side]) throw new Error("Unknown side");
  return (side + SPACE_SIDES / 2) % SPACE_SIDES || SPACE_SIDES;
};
