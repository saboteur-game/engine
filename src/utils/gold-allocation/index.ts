import { MAX_REWARD_CARDS, saboteurGoldAllocation } from "../../constants";
import Player from "../../models/player";
import GoldRewardCard from "../../models/cards/gold-reward-cards";
import RewardDeck from "../../models/reward-deck";
import { wrapIndexAt } from "../";

interface IGoldAllocation {
  [key: string]: GoldRewardCard[];
}

interface IPlayers {
  [key: string]: Player;
}

const getSuccessfulGoldDigger = (
  successfulPlayer: Player,
  playOrder: string[],
  players: IPlayers
): Player => {
  // If the saboteur reveals the gold finish card, the gold gets handed out
  // counter-clockwise starting from the dwarf who played last.
  // https://boardgamegeek.com/thread/82149/article/735273#735273
  if (successfulPlayer.isSaboteur) {
    const getIndex = wrapIndexAt(playOrder.length - 1);
    const startIndex = playOrder.indexOf(successfulPlayer.id);
    for (let i = 0; i < playOrder.length; i++) {
      const playerIndex = getIndex(startIndex - i);
      const nextPlayer = players[playOrder[playerIndex]];
      if (!nextPlayer.isSaboteur) return nextPlayer;
    }
  }

  return successfulPlayer;
};

const allocateGoldToGoldDiggers = (
  rewardDeck: RewardDeck,
  playOrder: string[],
  players: IPlayers,
  activePlayer: Player
): IGoldAllocation => {
  const goldAllocation: IGoldAllocation = {};

  const numberOfPlayers = playOrder.length;
  const numberOfRewardCards = Math.min(numberOfPlayers, MAX_REWARD_CARDS);
  const startingPlayer = getSuccessfulGoldDigger(
    activePlayer,
    playOrder,
    players
  );
  const goldDiggers = playOrder.filter(
    (playerId) => !players[playerId].isSaboteur
  );

  const startingPlayerIndex = goldDiggers.indexOf(startingPlayer.id);
  const getIndex = wrapIndexAt(goldDiggers.length - 1);
  for (let i = 0; i < numberOfRewardCards; i++) {
    const playerId = goldDiggers[getIndex(startingPlayerIndex - i)];
    const reward = rewardDeck.drawCard();
    if (reward) {
      goldAllocation[playerId] = (goldAllocation[playerId] || []).concat(
        reward
      );
    }
  }

  return goldAllocation;
};

const allocateGoldToSaboteurs = (
  rewardDeck: RewardDeck,
  playOrder: string[],
  players: IPlayers
): IGoldAllocation => {
  const goldAllocation: IGoldAllocation = {};

  const saboteurs = playOrder.filter(
    (playerId) => players[playerId].isSaboteur
  );
  const goldRewardAmount = saboteurGoldAllocation[saboteurs.length];
  saboteurs.forEach((playerId) => {
    goldAllocation[playerId] = (goldAllocation[playerId] || []).concat(
      rewardDeck.extractCardsToValue(goldRewardAmount)
    );
  });
  return goldAllocation;
};

export {
  IGoldAllocation,
  IPlayers,
  allocateGoldToGoldDiggers,
  allocateGoldToSaboteurs,
};
