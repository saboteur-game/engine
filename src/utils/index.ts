import { v4 as uuidv4 } from "uuid";

const generateId = (): string => uuidv4();

// https://stackoverflow.com/a/2450976
function shuffle<T>(originalArray: T[]): T[] {
  const array = [...originalArray];
  let currentIndex = array.length;
  let currentValue;
  let randomIndex;

  while (0 !== currentIndex) {
    randomIndex = Math.floor(Math.random() * currentIndex);
    currentIndex -= 1;
    currentValue = array[currentIndex];
    array[currentIndex] = array[randomIndex];
    array[randomIndex] = currentValue;
  }

  return array;
}

interface Pojo {
  [key: string]: unknown;
}

const multiply = <T>(length: number, getValue: () => T): T[] =>
  new Array(length).fill(undefined).map(() => getValue());

const randomBoolean = (): boolean => Math.random() < 0.5;

const wrapIndexAt = (max: number) => (index: number): number => {
  if (index > max) return index % max;
  if (index < 0) return max + (index % max);
  return index;
};

export { generateId, shuffle, multiply, randomBoolean, wrapIndexAt, Pojo };
