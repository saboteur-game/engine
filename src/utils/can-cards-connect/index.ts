import { Sides } from "../../models/cards/card";
import { TunnelCard } from "../../models/cards/path-cards";
import { getOppositeSide } from "../get-opposite-side";

function canCardsConnect(
  side: Sides,
  adjacentCard: TunnelCard | undefined,
  card: TunnelCard
): boolean {
  if (!adjacentCard) return true;

  const cardConnectors = card.connectors.map((connector) =>
    card.isUpsideDown ? getOppositeSide(connector) : connector
  );
  const adjacentCardConnectors = adjacentCard.connectors.map((connector) =>
    adjacentCard.isUpsideDown ? getOppositeSide(connector) : connector
  );

  const needsConnector = cardConnectors.includes(side);
  const hasConnector = adjacentCardConnectors.includes(getOppositeSide(side));
  return needsConnector === hasConnector;
}

export default canCardsConnect;
