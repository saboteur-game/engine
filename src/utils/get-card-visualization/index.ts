import { Sides } from "../../models/cards/card";
import { PathCard } from "../../models/cards/path-cards";
import { getOppositeSide } from "../get-opposite-side";

const getCardVisualization = (card: PathCard, center = "  "): string => {
  const getConnection = (passedSide: Sides): string => {
    const side = card.isUpsideDown ? getOppositeSide(passedSide) : passedSide;
    if (card.connectors.includes(side)) return "  ";
    return "██";
  };
  const top = `██${getConnection(Sides.top)}██`;
  const middle = `${getConnection(Sides.left)}${center}${getConnection(
    Sides.right
  )}`;
  const bottom = `██${getConnection(Sides.bottom)}██`;
  return `${top}\n${middle}\n${bottom}`;
};

export default getCardVisualization;
