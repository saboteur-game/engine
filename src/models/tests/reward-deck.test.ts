import { mocked } from "ts-jest/utils";
import RewardDeck from "../reward-deck";
import { shuffle } from "../../utils";
import GoldRewardCard from "../cards/gold-reward-cards";

const mockedShuffle = mocked(shuffle);

describe("RewardDeck", () => {
  let rewardDeck: RewardDeck;

  it("can be serialized", () => {
    expect(new RewardDeck().toJS()).toMatchSnapshot();
  });

  describe("drawCard", () => {
    describe("draw card when the pile is populated", () => {
      beforeEach(() => {
        rewardDeck = new RewardDeck();
      });

      it("draws a card from the deck and returns it", () => {
        expect(rewardDeck.getCardCount()).toBe(28);
        expect(rewardDeck.drawCard()).toMatchInlineSnapshot(`
          GoldRewardCard {
            "id": "test-id-1",
            "value": 3,
          }
        `);
        expect(rewardDeck.getCardCount()).toBe(27);
      });
    });

    describe("draw card when the pile is empty", () => {
      beforeEach(() => {
        mockedShuffle.mockReturnValue([]);
        rewardDeck = new RewardDeck();
      });

      it("returns undefined", () => {
        expect(rewardDeck.getCardCount()).toBe(0);
        expect(rewardDeck.drawCard()).toBe(undefined);
        expect(rewardDeck.getCardCount()).toBe(0);
      });
    });
  });

  describe("extractCardsToValue", () => {
    // e.g. cards [2,3,1,1] => to match 4 should return 2,1,1
    describe("when matching a combination of cards from the initial card", () => {
      beforeEach(() => {
        mockedShuffle.mockReturnValue([
          new GoldRewardCard(2),
          new GoldRewardCard(3),
          new GoldRewardCard(1),
          new GoldRewardCard(1),
          new GoldRewardCard(2),
        ]);
        rewardDeck = new RewardDeck();
      });

      it("draws cards up to a specific value", () => {
        expect(rewardDeck.getCardCount()).toBe(5);
        expect(rewardDeck.extractCardsToValue(4)).toMatchInlineSnapshot(`
          Array [
            GoldRewardCard {
              "id": "test-id-0",
              "value": 2,
            },
            GoldRewardCard {
              "id": "test-id-2",
              "value": 1,
            },
            GoldRewardCard {
              "id": "test-id-3",
              "value": 1,
            },
          ]
        `);
        expect(rewardDeck.getCardCount()).toBe(2);
      });
    });

    // e.g. cards [2,3,1,3] => to match 4 should return 3,1
    describe("when matching a combination of cards from not initial card", () => {
      beforeEach(() => {
        mockedShuffle.mockReturnValue([
          new GoldRewardCard(2),
          new GoldRewardCard(3),
          new GoldRewardCard(1),
          new GoldRewardCard(3),
        ]);
        rewardDeck = new RewardDeck();
      });

      it("draws cards up to a specific value", () => {
        expect(rewardDeck.getCardCount()).toBe(4);
        expect(rewardDeck.extractCardsToValue(4)).toMatchInlineSnapshot(`
          Array [
            GoldRewardCard {
              "id": "test-id-1",
              "value": 3,
            },
            GoldRewardCard {
              "id": "test-id-2",
              "value": 1,
            },
          ]
        `);
        expect(rewardDeck.getCardCount()).toBe(2);
      });
    });

    // e.g. cards [3,3,3,2] => to match 2 should return 2
    describe("when matching a single card which is last", () => {
      beforeEach(() => {
        mockedShuffle.mockReturnValue([
          new GoldRewardCard(3),
          new GoldRewardCard(3),
          new GoldRewardCard(3),
          new GoldRewardCard(2),
        ]);
        rewardDeck = new RewardDeck();
      });

      it("draws cards up to a specific value", () => {
        expect(rewardDeck.getCardCount()).toBe(4);
        expect(rewardDeck.extractCardsToValue(2)).toMatchInlineSnapshot(`
          Array [
            GoldRewardCard {
              "id": "test-id-3",
              "value": 2,
            },
          ]
        `);
        expect(rewardDeck.getCardCount()).toBe(3);
      });
    });

    // e.g. cards [2,3,3] => to match 4 should return 3,1 (1 being generated)
    describe("when matching a combination of cards which required generated gold", () => {
      beforeEach(() => {
        mockedShuffle.mockReturnValue([
          new GoldRewardCard(2),
          new GoldRewardCard(3),
          new GoldRewardCard(3),
        ]);
        rewardDeck = new RewardDeck();
      });

      it("draws cards up to a specific value", () => {
        expect(rewardDeck.getCardCount()).toBe(3);
        expect(rewardDeck.extractCardsToValue(4)).toMatchInlineSnapshot(`
          Array [
            GoldRewardCard {
              "id": "test-id-2",
              "value": 3,
            },
            GoldRewardCard {
              "id": "test-id-32",
              "value": 1,
            },
          ]
        `);
        expect(rewardDeck.getCardCount()).toBe(2);
      });
    });

    // e.g. cards [2,3,3] => to match 4 should return 3,1 (1 being generated)
    describe("when matching a combination of cards which requires all generated gold", () => {
      beforeEach(() => {
        mockedShuffle.mockReturnValue([
          new GoldRewardCard(3),
          new GoldRewardCard(3),
          new GoldRewardCard(3),
        ]);
        rewardDeck = new RewardDeck();
      });

      it("draws cards up to a specific value", () => {
        expect(rewardDeck.getCardCount()).toBe(3);
        expect(rewardDeck.extractCardsToValue(2)).toMatchInlineSnapshot(`
          Array [
            GoldRewardCard {
              "id": "test-id-32",
              "value": 1,
            },
            GoldRewardCard {
              "id": "test-id-33",
              "value": 1,
            },
          ]
        `);
        expect(rewardDeck.getCardCount()).toBe(3);
      });
    });

    // e.g. cards [2,3,1,2] => to match 4 should return 2,2
    describe("when matching a combination of cards which is spread across the deck", () => {
      beforeEach(() => {
        mockedShuffle.mockReturnValue([
          new GoldRewardCard(2),
          new GoldRewardCard(3),
          new GoldRewardCard(1),
          new GoldRewardCard(2),
        ]);
        rewardDeck = new RewardDeck();
      });

      it("draws cards up to a specific value", () => {
        expect(rewardDeck.getCardCount()).toBe(4);
        expect(rewardDeck.extractCardsToValue(4)).toMatchInlineSnapshot(`
          Array [
            GoldRewardCard {
              "id": "test-id-0",
              "value": 2,
            },
            GoldRewardCard {
              "id": "test-id-3",
              "value": 2,
            },
          ]
        `);
        expect(rewardDeck.getCardCount()).toBe(2);
      });
    });
  });
});
