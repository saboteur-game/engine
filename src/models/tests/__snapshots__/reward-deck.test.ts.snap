// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RewardDeck can be serialized 1`] = `
Object {
  "drawPile": Array [
    GoldRewardCard {
      "id": "test-id-1",
      "value": 3,
    },
    GoldRewardCard {
      "id": "test-id-2",
      "value": 3,
    },
    GoldRewardCard {
      "id": "test-id-3",
      "value": 3,
    },
    GoldRewardCard {
      "id": "test-id-4",
      "value": 3,
    },
    GoldRewardCard {
      "id": "test-id-5",
      "value": 2,
    },
    GoldRewardCard {
      "id": "test-id-6",
      "value": 2,
    },
    GoldRewardCard {
      "id": "test-id-7",
      "value": 2,
    },
    GoldRewardCard {
      "id": "test-id-8",
      "value": 2,
    },
    GoldRewardCard {
      "id": "test-id-9",
      "value": 2,
    },
    GoldRewardCard {
      "id": "test-id-10",
      "value": 2,
    },
    GoldRewardCard {
      "id": "test-id-11",
      "value": 2,
    },
    GoldRewardCard {
      "id": "test-id-12",
      "value": 2,
    },
    GoldRewardCard {
      "id": "test-id-13",
      "value": 1,
    },
    GoldRewardCard {
      "id": "test-id-14",
      "value": 1,
    },
    GoldRewardCard {
      "id": "test-id-15",
      "value": 1,
    },
    GoldRewardCard {
      "id": "test-id-16",
      "value": 1,
    },
    GoldRewardCard {
      "id": "test-id-17",
      "value": 1,
    },
    GoldRewardCard {
      "id": "test-id-18",
      "value": 1,
    },
    GoldRewardCard {
      "id": "test-id-19",
      "value": 1,
    },
    GoldRewardCard {
      "id": "test-id-20",
      "value": 1,
    },
    GoldRewardCard {
      "id": "test-id-21",
      "value": 1,
    },
    GoldRewardCard {
      "id": "test-id-22",
      "value": 1,
    },
    GoldRewardCard {
      "id": "test-id-23",
      "value": 1,
    },
    GoldRewardCard {
      "id": "test-id-24",
      "value": 1,
    },
    GoldRewardCard {
      "id": "test-id-25",
      "value": 1,
    },
    GoldRewardCard {
      "id": "test-id-26",
      "value": 1,
    },
    GoldRewardCard {
      "id": "test-id-27",
      "value": 1,
    },
    GoldRewardCard {
      "id": "test-id-28",
      "value": 1,
    },
  ],
  "id": "test-id-0",
}
`;
